pex/variables.py

```diff
@@ -668,45 +672,118 @@ def unzip_dir(
+    has_interpreter_constraints,  # type: bool
+    interpreter=None,  # type: Optional[PythonInterpreter]
     pex_path=None,  # type: Optional[str]
 ):
     # type: (...) -> str
-    # The venv contents are affected by which interpreter is selected as well as which PEX files are
-    # in play. The former can be influenced by interpreter constraints changes as well as PEX_PYTHON
-    # and PEX_PYTHON_PATH. The latter is influenced via PEX_PATH.
-    venv_contents = {
-        "interpreter_constraints": sorted(interpreter_constraints),
-        "PEX_PYTHON": ENV.PEX_PYTHON,
-        "PEX_PYTHON_PATH": ENV.PEX_PYTHON_PATH,
-        "strip_pex_env": strip_pex_env,
-        "pex_path": {},
-    }  # type: Dict[str, Any]
+
+    # The venv contents are affected by which PEX files are in play as well as which interpreter
+    # is selected. The former is influenced via PEX_PATH and the is influenced by interpreter
+    # constraints, PEX_PYTHON and PEX_PYTHON_PATH.
+
+    pex_path_contents = {}  # type: Dict[str, Dict[str, str]]
+    venv_contents = {"pex_path": pex_path_contents}  # type: Dict[str, Any]
 
     # PexInfo.pex_path and PEX_PATH are merged by PEX at runtime, so we include both and hash just
     # the distributions since those are the only items used from PEX_PATH adjoined PEX files; i.e.:
     # neither the entry_point nor any other PEX file data or metadata is used.
-    def add_pex_path_items(pex_path):
-        if not pex_path:
+    def add_pex_path_items(path):
+        # type: (Optional[str]) -> None
+        if not path:
             return
         from pex.pex_info import PexInfo
 
-        pex_path_contents = venv_contents["pex_path"]
-        for pex in pex_path.split(":"):
+        for pex in path.split(":"):
             pex_path_contents[pex] = PexInfo.from_pex(pex).distributions
 
     add_pex_path_items(pex_path)
     add_pex_path_items(ENV.PEX_PATH)
 
+    # There are two broad cases of interpreter selection, Pex chooses and user chooses. For the
+    # former we hash Pex's selection criteria and for the latter we write down the actual
+    # interpreter the user selected. N.B.: The pex_hash already includes an encoding of the
+    # interpreter constraints, if any; so we don't hash those here again for the Pex chooses case.
+
+    # This is relevant in all cases of interpreter selection and restricts the valid search path
+    # for interpreter constraints, PEX_PYTHON and otherwise unconstrained PEXes.
+    venv_contents["PEX_PYTHON_PATH"] = ENV.PEX_PYTHON_PATH
+
+    interpreter_path = None  # type: Optional[str]
+    imprecise_pex_python = ENV.PEX_PYTHON and not os.path.isabs(ENV.PEX_PYTHON)
+    if imprecise_pex_python:
+        # For PEX_PYTHON=python3.7, for example, we just write down the constraint and let the PEX
+        # runtime determine the path of an appropriate python3.7, if any (Pex chooses).
+        venv_contents["PEX_PYTHON"] = ENV.PEX_PYTHON
+    elif not has_interpreter_constraints:
```

Comment:
If there is a `PEX_PYTHON`, can't you assume (based on the errors in #1427) that it will be used here, regardless of whether there are constraints? It might error if the constraints mismatch, but...

Reply:
That sounds right. I need to think on this though. The logic is getting pretty intricate.

Reply:
Alright. Definitely correct. Thanks for the close look. Fixed.